src/bootstrap/builder.rs

```diff
@@ -168,33 +168,56 @@ impl PathSet {
         PathSet::Set(BTreeSet::new())
     }
 
-    fn one<P: Into<PathBuf>>(path: P, kind: Kind) -> PathSet {
-        let mut set = BTreeSet::new();
-        set.insert(TaskPath { path: path.into(), kind: Some(kind) });
-        PathSet::Set(set)
-    }
-
-    fn has(&self, needle: &Path, module: Option<Kind>) -> bool {
-        let check = |p: &TaskPath| {
-            if let (Some(p_kind), Some(kind)) = (&p.kind, module) {
-                p.path.ends_with(needle) && *p_kind == kind
-            } else {
-                p.path.ends_with(needle)
+    /// Return all `TaskPath`s in `Self` that contain any of the `needles`.
+    ///
+    /// This is used for `StepDescription::krate`, which passes all matching crates at once to
+    /// `Step::make_run`, rather than calling it many times with a single crate.
+    fn intersection(&self, needles: &mut Vec<&Path>, module: Option<Kind>) -> PathSet {
+        let mut check = |p| {
+            for (i, n) in needles.iter().enumerate() {
+                let matched = Self::check(p, n, module);
+                if matched {
+                    needles.remove(i);
+                    return true;
+                }
             }
+            false
         };
+        match self {
+            PathSet::Set(set) => PathSet::Set(set.iter().filter(|&p| check(p)).cloned().collect()),
+            PathSet::Suite(suite) => {
+                if check(suite) {
+                    self.clone()
+                } else {
+                    PathSet::empty()
+                }
+            }
+        }
+    }
 
+    fn has(&self, needle: &Path, module: Option<Kind>) -> bool {
         match self {
-            PathSet::Set(set) => set.iter().any(check),
-            PathSet::Suite(suite) => check(suite),
+            PathSet::Set(set) => set.iter().any(|p| Self::check(p, needle, module)),
+            PathSet::Suite(suite) => Self::check(suite, needle, module),
+        }
+    }
+
+    // internal use only
+    fn check(p: &TaskPath, needle: &Path, module: Option<Kind>) -> bool {
+        if let (Some(p_kind), Some(kind)) = (&p.kind, module) {
+            p.path.ends_with(needle) && *p_kind == kind
+        } else {
+            p.path.ends_with(needle)
         }
     }
 
-    fn path(&self, builder: &Builder<'_>) -> PathBuf {
+    fn paths(&self) -> Vec<PathBuf> {
         match self {
             PathSet::Set(set) => {
-                set.iter().next().map(|p| &p.path).unwrap_or(&builder.build.src).clone()
+                // NOTE: empty set is already handled in StepDescription::run below
+                set.iter().map(|p| p.path.clone()).collect()
```

Comment:
This seems a little iffy to me -- I would have expected a single path in the output for a set on input. IOW, if we have e.g. set=[library/std, library/core, ...] then on output it probably makes sense to either keep just the first entry or have the vec here be a list of PathSets, not PathBufs.

In general the whole point of the PathSet abstraction was to let us have aliases without running multiple times. I suspect that krate -> all_krates is potentially the right move here, which puts all the dependencies into separate PathSets IIRC.

Reply:
I don't quite follow why you expect a single path. Here is how `paths` is used currently:
```rust
        for target in targets {
            let run = RunConfig { builder, paths: pathset.paths(), target: *target };
            (self.make_run)(run);
        }
```
and here is an example of multiple paths being used:
```rust
    fn make_run(run: RunConfig<'_>) {
        let builder = run.builder;
        let host = run.build_triple();
        let compiler = builder.compiler_for(builder.top_stage, host, host);
        let crates = run.paths.iter().map(|p| builder.crate_paths[p].clone()).collect();
        let test_kind = builder.kind.into();

        builder.ensure(CrateLibrustc { compiler, target: run.target, test_kind, crates });
    }
```
If we use a list of `PathSet`s instead, that defeats the whole point of this change, which is to run that step only once with all relevant paths.

Reply:
> I suspect that krate -> all_krates is potentially the right move here, which puts all the dependencies into separate PathSets IIRC.

I actually want to get rid of `all_krates` completely, eventually, because it makes it impossible to do steps for a single crate, you *have* to do the whole step at once because you can't distinguish which one should be run.

Reply:
> I actually want to get rid of all_krates completely, eventually, because it makes it impossible to do steps for a single crate, you have to do the whole step at once because you can't distinguish which one should be run.

I don't understand this -- maybe all_krates is confusing here.

To take a concrete step, let's call CrateLibrary the step for testing `library/{std,alloc,core}` (others too but for simplicity let's assume just these). I would expect that to configure (whether through all_krates or through some other mechanism) the  set of paths to have 3 separate pathsets (inventing aliases here, we may not actually want these particular ones): `{std, library/std}, {alloc, library/alloc}, {core, library/core}`.

If running by default (x.py test, no args), we'd want to invoke make_run with a vec of all three pathsets (since we have nothing more specific) (`vec![{std, library/std}, {alloc, library/alloc}, {core, library/core}]`), this should configure crates to be `[core, std, alloc]`.
For x.py test core library/std, we'd pass `vec![{std, library/std}, {core, library/core}]`.
For `x.py test --exclude core`, we'd pass `vec![{std, library/std}, {alloc, library/alloc}]`.

Does that make sense? Am I missing something?

It may be that passing the PathSet into run_make isn't *really* any better than just passing one of the elements, but in practice I think it's the right thing to do. In the future we might want PathSet to really represent something with an ID or whatever to make it easier to have the steps take Cargo-level metadata (crate IDs) -> human paths (for CLI invocation) -> Cargo-level metadata (crate IDs) in a way that round trips reliably and doesn't require as much guessing when looking things up in hashmaps.

But that can come later.

Reply:
> If running by default (x.py test, no args), we'd want to invoke make_run with a vec of all three pathsets (since we have nothing more specific) (vec![{std, library/std}, {alloc, library/alloc}, {core, library/core}]), this should configure crates to be [core, std, alloc].

ahh, I see - the "set" here is all possible aliases for the same crate (not the full Step - `all_krates` registers all aliases, but only with `Step` granularity).

I guess that doesn't hurt, but I'm not sure why it would be related to this PR? If you look at the `path` function I deleted, it also uses PathBuf and not PathSet, because nothing needs to know all possible aliases currently. I am kind of worried this overcomplicates the API - I would rather not build PathSet into this until and unless we end up using your ID scheme you mentioned in the last paragraph.

That said, it wouldn't be *that* much more complicated, so I'm ok with using PathSet instead if you feel strongly. But it would introduce additional work into each step having to go from the set to the path; and in most cases I think it will end up with `set.first().unwrap()` or similar which makes me a little nervous.

Reply:
I think the line I am commenting on previously retrieved the first element of the pathset, but it looks like it now takes all of the elements and pulls them in. That seems wrong to me; we should never flatten a PathSet (e.g., [{core, library/core}, {std, library/std}] -> [core, library/core, std, library/std]). Maybe I'm misreading the code though?

Reply:
ah! ok, I see what's going on - for `PathSet::one`, like `krate` uses, this will be correct, but for `all_krates` the behavior is wrong, because it flattens aliases (and same for `paths`).

I don't think any step that currently uses `all_crates` or `paths` actually uses `run.path` and I worry a little about changing the API too much to account for them. But I agree flattening is wrong - I will change the steps that actually use `run.path` to unwrap the path out of the pathset.

Reply:
Hm. So I think the rationale I'd expect is that we'd invoke a given Step once per PathSet, which I think corresponds to the new behavior (i.e., with run.paths having a full list). We'd still only actually run it once for most steps today since we cache steps (i.e., we invoke make_run multiple times but the step body is only run once).

I would expect it to be a hard requirement that we can work with multiple invocations of the same step across pathsets for things like x.py test library/std library/test to work, so that does seem like a blocker if we can't support that with this PR. If it's just the not-yet-"good" mode steps that are in all-crates-aliased-together mode that don't support it though that seems OK, we can largely ignore the issue as it's not really expected to come up in practice. I think adding just another line or something saying "or a previous rule already matched that step" would be fine there.

Reply:
> I would expect it to be a hard requirement that we can work with multiple invocations of the same step across pathsets for things like x.py test library/std library/test to work

Just to clarify, std and test are in the same pathset, that already works today. The difference between pathsets is basically "one pathset per function on `should_run`.

> We'd still only actually run it once for most steps today since we cache steps. If it's just the not-yet-"good" mode steps that are in all-crates-aliased-together mode that don't support it though that seems OK.

👍 Makes sense, I think this should be possible to get working.

Reply:
> Just to clarify, std and test are in the same pathset, that already works today.

Hm, I would not expect them to share a pathset for the test steps I think, since IIRC we support testing *just* std or just test (meaning two different tasks and so they should be in separate pathsets).